src/backend/x11/mod.rs

```diff
@@ -0,0 +1,921 @@
+
+        // Next find a visual using the supported depth
+        let visual_id = depth
+            .visuals
+            .iter()
+            .find(|visual| visual.class == VisualClass::TRUE_COLOR)
+            .ok_or(CreateWindowError::NoVisual)?
+            .visual_id;
+
+        let format = match depth.depth {
+            24 => DrmFourcc::Xrgb8888,
+            32 => DrmFourcc::Argb8888,
+            _ => unreachable!(),
+        };
+
+        // Make a colormap
+        let colormap = connection.generate_id()?;
+        connection.create_colormap(ColormapAlloc::NONE, colormap, screen.root, visual_id)?;
+
+        let atoms = Atoms::new(&*connection)?.reply()?;
+
+        let window = Arc::new(WindowInner::new(
+            Arc::downgrade(&connection),
+            screen,
+            properties,
+            atoms,
+            depth.clone(),
+            visual_id,
+            colormap,
+        )?);
+
+        let source = X11Source::new(
+            connection.clone(),
+            window.id,
+            atoms._SMITHAY_X11_BACKEND_CLOSE,
+            logger.clone(),
+        );
+
+        info!(logger, "Window created");
+
+        let (resize_send, resize_recv) = mpsc::channel();
+
+        let backend = X11Backend {
+            log: logger,
+            source,
+            connection,
+            window,
+            key_counter: Arc::new(AtomicU32::new(0)),
+            depth,
+            visual_id,
+            screen_number,
+            resize: resize_send,
+        };
+
+        let surface = X11Surface::new(&backend, format, resize_recv)?;
+
+        Ok((backend, surface))
+    }
+
+    /// Returns the default screen number of the X server.
+    pub fn screen(&self) -> usize {
+        self.screen_number
+    }
+
+    /// Returns the underlying connection to the X server.
+    pub fn connection(&self) -> &RustConnection {
+        &*self.connection
+    }
+
+    /// Returns a handle to the X11 window this input backend handles inputs for.
+    pub fn window(&self) -> Window {
+        self.window.clone().into()
+    }
+}
+
+/// An X11 surface which uses GBM to allocate and present buffers.
+#[derive(Debug)]
+pub struct X11Surface {
+    connection: Weak<RustConnection>,
+    window: Window,
+    resize: Receiver<Size<u16, Logical>>,
+    device: gbm::Device<RawFd>,
+    format: DrmFourcc,
+    width: u16,
+    height: u16,
+    current: Dmabuf,
+    next: Dmabuf,
+}
+
+impl X11Surface {
+    fn new(
+        backend: &X11Backend,
+        format: DrmFourcc,
+        resize: Receiver<Size<u16, Logical>>,
+    ) -> Result<X11Surface, X11Error> {
+        let connection = &backend.connection;
+        let window = backend.window();
+
+        // Determine which drm-device the Display is using.
+        let screen = &connection.setup().roots[backend.screen()];
+        let dri3 = connection.dri3_open(screen.root, 0)?.reply()?;
+
+        let drm_device_fd = dri3.device_fd;
+        // Duplicate the drm_device_fd
```

Comment:
Is it necessary to duplicate the fd? :-(

I would hope that since you own the struct, you can just move members out of it. In fact, doesn't `let drm_device_fd = dri3.device_fd;` already move the FD out? Can't you replace the `fcntl(DUPFD)` call with `dri3.device_fd.into_raw_fd()`

Reply:
See comment regarding creation of gbm::Device